File: platform/polyglot/rag_llm_synthesis/mjs/src/json-extractor.mjs
/**
 * @fileoverview Extract JSON from LLM responses.
 */

/**
 * Extract JSON from LLM response, stripping markdown fences if present.
 *
 * @param {string} text - Raw LLM response text
 * @returns {string} Cleaned JSON string, or original text if no valid JSON found
 */
export function extractJson(text) {
  let stripped = text.trim();

  // Remove markdown code fences
  if (stripped.startsWith('```')) {
    const lines = stripped.split('\n');
    // Drop first line (```json) and last line (```)
    const inner = lines.slice(1);
    if (inner.length && inner[inner.length - 1].trim() === '```') {
      inner.pop();
    }
    stripped = inner.join('\n').trim();
  }

  // Validate it's parseable JSON
  try {
    JSON.parse(stripped);
    return stripped;
  } catch {
    // Try to find a JSON object in the text
    const match = stripped.match(/\{[\s\S]*\}/);
    if (match) {
      try {
        JSON.parse(match[0]);
        return match[0];
      } catch {
        // fall through
      }
    }
    return text;
  }
}

File: platform/polyglot/rag_llm_synthesis/py/rag_llm_synthesis/json_extractor.py
"""Extract JSON from LLM responses."""

import json
import re


def extract_json(text: str) -> str:
    """Extract JSON from LLM response, stripping markdown fences if present.

    Args:
        text: Raw LLM response text that may contain JSON wrapped in markdown fences.

    Returns:
        Cleaned JSON string. Returns original text if no valid JSON found.
    """
    stripped = text.strip()
    # Remove markdown code fences
    if stripped.startswith("```"):
        lines = stripped.split("\n")
        # Drop first line (```json) and last line (```)
        lines = lines[1:]
        if lines and lines[-1].strip() == "```":
            lines = lines[:-1]
        stripped = "\n".join(lines).strip()
    # Validate it's parseable JSON
    try:
        json.loads(stripped)
        return stripped
    except json.JSONDecodeError:
        # Try to find a JSON object in the text
        match = re.search(r'\{[\s\S]*\}', stripped)
        if match:
            try:
                json.loads(match.group())
                return match.group()
            except json.JSONDecodeError:
                pass
        return text
